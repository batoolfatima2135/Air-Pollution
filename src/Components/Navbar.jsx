import { NavLink, useLocation } from 'react-router-dom';
import settings from '../Assets/settings.png';
import mic from '../Assets/mic.png';
import back from '../Assets/back.png';

const Navbar = () => {
  const location = useLocation();
  const currentUrl = location.pathname;
  return (
    <nav className="navbar navbar-expand-lg navbar-expand-sm mx-2 bg-dark">
      {currentUrl.includes('/details') ? <NavLink to="/"><img src={back} alt="back" height="20px" /></NavLink> : null}
      <h3 className="text-light m-0">Air pollution Data</h3>
      <ul className="navbar-nav ms-auto d-flex flex-row">
        <li className="nav-item">
          <NavLink to="/" className="nav-link">
            <img src={mic} alt="mic" height="30px" />
          </NavLink>
        </li>
        <li className="nav-item">
          <NavLink to="/" className="nav-link">
            <img src={settings} alt="setting" />
          </NavLink>
        </li>
      </ul>
    </nav>
  );
};
export default Navbar;
