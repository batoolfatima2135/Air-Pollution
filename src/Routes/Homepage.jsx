import { useState, useMemo } from 'react';
import { useSelector } from 'react-redux';
import { NavLink } from 'react-router-dom';
import arrow from '../Assets/arrow.png';
import map from '../Assets/map.png';

const Homepage = () => {
  const [searchInput, setSearchInput] = useState('');
  const [selectedCountry, setSelectedCountry] = useState('ALL');
  const [selectedAQI, setSelectedAQI] = useState('ALL');
  const [heading, setHeading] = useState('CITIES');
  const { data, isLoading } = useSelector((state) => state.airPollutionData);
  const uniqueCountries = [];
  const uniqueAQI = [];
  const filteredData = useMemo(() => {
    let result = data;

    if (searchInput !== '') {
      result = result.filter((item) => item.name.toLowerCase().includes(searchInput.toLowerCase()));
    } else {
      if (selectedCountry !== 'ALL') {
        result = result.filter((item) => item.country === selectedCountry);
      }

      if (selectedAQI !== 'ALL') {
        result = result.filter((item) => item.aqi.toString() === selectedAQI);
      }
    }

    return result;
  }, [data, searchInput, selectedCountry, selectedAQI]);

  const filteredCountries = data.filter((item) => {
    if (!uniqueCountries.includes(item.country)) {
      uniqueCountries.push(item.country);
      return true;
    }
    return false;
  });
  const filteredAQI = data.filter((item) => {
    if (!uniqueAQI.includes(item.aqi)) {
      uniqueAQI.push(item.aqi);
      return true;
    }
    return false;
  });
  const handleChange = (event) => {
    const { value } = event.target;
    setSearchInput(value);
    if (!value === '') {
      setHeading(`Search results for "${value}"`);
    }
  };

  const handleCountryChange = (event) => {
    const { value } = event.target;
    setSelectedCountry(value);
    if (selectedAQI !== 'ALL' && value !== 'ALL') {
      setHeading(`CITIES IN ${value} WITH AQI ${selectedAQI}`);
    } else {
      setHeading(`CITIES IN ${value}`);
    }
  };

  const handleAQIChange = (event) => {
    const { value } = event.target;
    setSelectedAQI(value);
    if (selectedCountry !== 'ALL' && value !== 'ALL') {
      setHeading(`CITIES IN ${selectedCountry} WITH AQI ${value}`);
    } else {
      setHeading(`CITIES WITH AQI ${value}`);
    }
  };

  const getEmojiFace = (value) => {
    switch (value) {
      case 2:
        return '😊'; // Smiling face for value 1
      case 1:
        return '😄'; // Grinning face for value 2
      case 3:
        return '😐'; // Neutral face for value 3
      case 4:
        return '😕'; // Confused face for value 4
      case 5:
        return '😢'; // Crying face for value 5
      default:
        return '❓'; // A question mark emoji for unknown values
    }
  };
  return (
    <div>
      <header>
        <div className="row m-0 p-0 ">
          <div className="col-12 p-3 text-center">
            <div className="row justify-content-center">
              <div className="col-12">
                <div className="input-group-sm d-flex">
                  <input
                    type="text"
                    className="form-control rounded-0"
                    placeholder="Search by city"
                    onChange={handleChange}
                    value={searchInput}
                  />
                  <div className="col-3">
                    <select
                      className="form-select rounded-0 form-select-sm "
                      aria-label="Select"
                      value={selectedCountry}
                      onChange={handleCountryChange}
                    >
                      <option value="ALL">Country</option>
                      {filteredCountries.map((item) => (
                        <option value={item.country} key={item.id}>{item.country}</option>
                      ))}
                    </select>
                  </div>
                  <div className="col-3">
                    <select
                      className="form-select form-select-sm  rounded-0"
                      aria-label="Select"
                      value={selectedAQI}
                      onChange={handleAQIChange}
                    >
                      <option value="ALL">AQI</option>
                      {filteredAQI.sort((a, b) => a.aqi - b.aqi).map((item) => (
                        <option value={item.aqi} key={item.id}>{item.aqi}</option>
                      ))}
                    </select>
                  </div>
                </div>
              </div>
            </div>
            <h5 className="text-light mt-4"><b>{heading}</b></h5>
          </div>
        </div>
      </header>
      <div className="row bg-dark p-2 w-100">
        <p className="text-light mx-2 m-0"><b>AQI INDEX</b></p>
      </div>
      <div className="row m-0">
        {!isLoading && filteredData.length > 0 ? filteredData.map((item, index) => (
          <div className={`col-6  ${(index + 1) % 4 === 2 || (index + 1) % 4 === 3 ? 'bg-dark' : 'bg-primary'} px-3`} key={item.id}>
            <NavLink to={`/details/${item.id}`} className="text-light text-decoration-none">
              <img src={arrow} alt="arrow" height="20px" className="float-end my-1" />
              <img src={map} alt="map" className="img-fluid" />
              <div className="row justify-content-end align-items-end">
                <div className="col-12 text-end">
                  <p className="text-uppercase m-0"><b>{item.name}</b></p>
                  <p className="m-0">{`${item.state}, ${item.country}`}</p>
                  <p>
                    <b>{`AQI : ${item.aqi}`}</b>
                    {` ${getEmojiFace(item.aqi)}`}
                  </p>
                </div>
              </div>
            </NavLink>
          </div>

        )) : <div className="p-4 pb-0 text-center text-light"><h5 className="m-0 p-0">Nothing found</h5></div>}
        {isLoading && (
          <div className="p-4 text-center text-light"><h5>Loading...</h5></div>
        )}
      </div>
    </div>

  );
};
export default Homepage;
