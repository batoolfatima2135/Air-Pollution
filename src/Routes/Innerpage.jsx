import React from 'react';
import { useSelector } from 'react-redux';
import { useParams } from 'react-router-dom';
import PollutantGraph from '../Components/Pollutantgraph';
import map from '../Assets/map.png';

const Innerpage = () => {
  const data = useSelector((state) => state.airPollutionData.data);
  const { id } = useParams();
  const item = data.find((item) => item.id === id);
  const getEmojiFace = (value) => {
    switch (value) {
      case 2:
        return '😊'; // Smiling face for value 1
      case 1:
        return '😄'; // Grinning face for value 2
      case 3:
        return '😐'; // Neutral face for value 3
      case 4:
        return '😕'; // Confused face for value 4
      case 5:
        return '😢'; // Crying face for value 5
      default:
        return '❓'; // A question mark emoji for unknown values
    }
  };
  return (
    <>
      <div className="py-5 m-0 row justify-content-center" style={{ backgroundImage: `url(${map})`, backgroundSize: 'cover' }}>
        <div className="col-10 text-center text-light">
          <h1>
            {item.name}
            {` ${getEmojiFace(item.aqi)}`}
          </h1>
          <p className="m-0"><b>{`${item.state}, ${item.country}`}</b></p>
          <p className="m-0">
            <b>
              {`AQI: ${item.aqi}`}

            </b>
          </p>
        </div>

      </div>
      <div className="row my-3 row justify-content-center">
        <h3 className="text-light text-center bg-dark  p-2">GRAPH</h3>
        <PollutantGraph components={item.components} />
      </div>
      <div className="row justify-content-center">
        <h3 className="text-light bg-dark p-2 text-center">DATA</h3>
        <div className="col-11 text-center">
          <table className="table table-bordered border-primary">
            <thead>
              <tr>
                <th scope="col" className="text-start text-dark">POLLUTANT</th>
                <th scope="col" className="text-dark">LEVEL</th>
              </tr>
            </thead>
            <tbody>
              {Object.entries(item.components).map(([pollutant, level]) => (
                <tr key={level}>
                  <td className="text-start">{pollutant}</td>
                  <td>{level}</td>
                </tr>
              ))}
            </tbody>
          </table>
        </div>

      </div>
    </>

  );
};
export default Innerpage;
