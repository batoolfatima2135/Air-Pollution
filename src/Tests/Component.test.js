import { render } from '@testing-library/react';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import {
  RouterProvider, MemoryRouter, createMemoryRouter, BrowserRouter as Router,
} from 'react-router-dom';
import store from '../Redux/Store';
import Navbar from '../Components/Navbar';
import Innerpage from '../Routes/Innerpage';
import Homepage from '../Routes/Homepage';
import PollutantGraph from '../Components/Pollutantgraph';

const mockStore = configureStore([]);

describe('Component testing', () => {
  test('Navbar component should render correctly', () => {
    const { asFragment } = render(
      <Provider store={store}>
        <Router>
          <Navbar />
        </Router>
      </Provider>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
  test('Homepage should render correctly', () => {
    const mockState = {
      airPollutionData: {
        data: [
          {
            id: '1', name: 'City 1', state: 'State 1', country: 'Country 1', aqi: 3, components: {},
          },
          {
            id: '2', name: 'City 2', state: 'State 2', country: 'Country 2', aqi: 4, components: {},
          },
        ],
        isLoading: false,
      },
    };

    const store = mockStore(mockState);

    const { asFragment } = render(
      <Provider store={store}>
        <MemoryRouter>
          <Homepage />
        </MemoryRouter>
      </Provider>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
  test('Innerpage should render correctly', () => {
    const routes = [
      {
        path: '/:id',
        element: <Innerpage />,
      },
    ];

    const router = createMemoryRouter(routes, {
      initialEntries: ['/', '/ADBE'],
      initialIndex: 1,
    });

    const { asFragment } = render(
      <Provider store={store}>
        <RouterProvider router={router} />
      </Provider>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
describe('PollutantGraph', () => {
  const mockComponents = {
    co: 10,
    no2: 20,
    o3: 30,
    pm10: 40,
  };

  it('should render the PollutantGraph component correctly', () => {
    render(<PollutantGraph components={mockComponents} />);

    // Your assertions
  });
});
