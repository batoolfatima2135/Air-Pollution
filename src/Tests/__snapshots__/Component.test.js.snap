// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component testing Homepage should render correctly 1`] = `
<DocumentFragment>
  <div>
    <header>
      <div
        class="row m-0 p-0 "
      >
        <div
          class="col-12 p-3 text-center"
        >
          <div
            class="row justify-content-center"
          >
            <div
              class="col-12"
            >
              <div
                class="input-group-sm d-flex"
              >
                <input
                  class="form-control rounded-0"
                  placeholder="Search by city"
                  type="text"
                  value=""
                />
                <div
                  class="col-3"
                >
                  <select
                    aria-label="Select"
                    class="form-select rounded-0 form-select-sm "
                  >
                    <option
                      value="ALL"
                    >
                      Country
                    </option>
                    <option
                      value="Country 1"
                    >
                      Country 1
                    </option>
                    <option
                      value="Country 2"
                    >
                      Country 2
                    </option>
                  </select>
                </div>
                <div
                  class="col-3"
                >
                  <select
                    aria-label="Select"
                    class="form-select form-select-sm  rounded-0"
                  >
                    <option
                      value="ALL"
                    >
                      AQI
                    </option>
                    <option
                      value="3"
                    >
                      3
                    </option>
                    <option
                      value="4"
                    >
                      4
                    </option>
                  </select>
                </div>
              </div>
            </div>
          </div>
          <h5
            class="text-light mt-4"
          >
            <b>
              CITIES
            </b>
          </h5>
        </div>
      </div>
    </header>
    <div
      class="row bg-dark p-2 w-100"
    >
      <p
        class="text-light mx-2 m-0"
      >
        <b>
          AQI INDEX
        </b>
      </p>
    </div>
    <div
      class="row m-0"
    >
      <div
        class="col-6  bg-primary px-3"
      >
        <a
          class="text-light text-decoration-none"
          href="/details/1"
        >
          <img
            alt="arrow"
            class="float-end my-1"
            height="20px"
            src="arrow.png"
          />
          <img
            alt="map"
            class="img-fluid"
            src="map.png"
          />
          <div
            class="row justify-content-end align-items-end"
          >
            <div
              class="col-12 text-end"
            >
              <p
                class="text-uppercase m-0"
              >
                <b>
                  City 1
                </b>
              </p>
              <p
                class="m-0"
              >
                State 1, Country 1
              </p>
              <p>
                <b>
                  AQI : 3
                </b>
                 üòê
              </p>
            </div>
          </div>
        </a>
      </div>
      <div
        class="col-6  bg-dark px-3"
      >
        <a
          class="text-light text-decoration-none"
          href="/details/2"
        >
          <img
            alt="arrow"
            class="float-end my-1"
            height="20px"
            src="arrow.png"
          />
          <img
            alt="map"
            class="img-fluid"
            src="map.png"
          />
          <div
            class="row justify-content-end align-items-end"
          >
            <div
              class="col-12 text-end"
            >
              <p
                class="text-uppercase m-0"
              >
                <b>
                  City 2
                </b>
              </p>
              <p
                class="m-0"
              >
                State 2, Country 2
              </p>
              <p>
                <b>
                  AQI : 4
                </b>
                 üòï
              </p>
            </div>
          </div>
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Component testing Innerpage should render correctly 1`] = `
<DocumentFragment>
  <h2>
    Unexpected Application Error!
  </h2>
  <h3
    style="font-style: italic;"
  >
    Cannot read properties of undefined (reading 'name')
  </h3>
  <pre
    style="padding: 0.5rem; background-color: rgba(200, 200, 200, 0.5);"
  >
    TypeError: Cannot read properties of undefined (reading 'name')
    at Innerpage (C:\\Users\\LENOVO\\Desktop\\air-pollution-data\\src\\Routes\\Innerpage.jsx:32:19)
    at renderWithHooks (C:\\Users\\LENOVO\\Desktop\\air-pollution-data\\node_modules\\react-dom\\cjs\\react-dom.development.js:16305:18)
    at mountIndeterminateComponent (C:\\Users\\LENOVO\\Desktop\\air-pollution-data\\node_modules\\react-dom\\cjs\\react-dom.development.js:20074:13)
    at beginWork (C:\\Users\\LENOVO\\Desktop\\air-pollution-data\\node_modules\\react-dom\\cjs\\react-dom.development.js:21587:16)
    at beginWork$1 (C:\\Users\\LENOVO\\Desktop\\air-pollution-data\\node_modules\\react-dom\\cjs\\react-dom.development.js:27426:14)
    at performUnitOfWork (C:\\Users\\LENOVO\\Desktop\\air-pollution-data\\node_modules\\react-dom\\cjs\\react-dom.development.js:26560:12)
    at workLoopSync (C:\\Users\\LENOVO\\Desktop\\air-pollution-data\\node_modules\\react-dom\\cjs\\react-dom.development.js:26466:5)
    at renderRootSync (C:\\Users\\LENOVO\\Desktop\\air-pollution-data\\node_modules\\react-dom\\cjs\\react-dom.development.js:26434:7)
    at recoverFromConcurrentError (C:\\Users\\LENOVO\\Desktop\\air-pollution-data\\node_modules\\react-dom\\cjs\\react-dom.development.js:25850:20)
    at performConcurrentWorkOnRoot (C:\\Users\\LENOVO\\Desktop\\air-pollution-data\\node_modules\\react-dom\\cjs\\react-dom.development.js:25750:22)
    at flushActQueue (C:\\Users\\LENOVO\\Desktop\\air-pollution-data\\node_modules\\react\\cjs\\react.development.js:2667:24)
    at act (C:\\Users\\LENOVO\\Desktop\\air-pollution-data\\node_modules\\react\\cjs\\react.development.js:2582:11)
    at C:\\Users\\LENOVO\\Desktop\\air-pollution-data\\node_modules\\@testing-library\\react\\dist\\act-compat.js:46:25
    at renderRoot (C:\\Users\\LENOVO\\Desktop\\air-pollution-data\\node_modules\\@testing-library\\react\\dist\\pure.js:161:26)
    at render (C:\\Users\\LENOVO\\Desktop\\air-pollution-data\\node_modules\\@testing-library\\react\\dist\\pure.js:247:10)
    at Object.&lt;anonymous&gt; (C:\\Users\\LENOVO\\Desktop\\air-pollution-data\\src\\Tests\\Component.test.js:64:34)
    at Promise.then.completed (C:\\Users\\LENOVO\\Desktop\\air-pollution-data\\node_modules\\jest-circus\\build\\utils.js:391:28)
    at new Promise (&lt;anonymous&gt;)
    at callAsyncCircusFn (C:\\Users\\LENOVO\\Desktop\\air-pollution-data\\node_modules\\jest-circus\\build\\utils.js:316:10)
    at _callCircusTest (C:\\Users\\LENOVO\\Desktop\\air-pollution-data\\node_modules\\jest-circus\\build\\run.js:218:40)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at _runTest (C:\\Users\\LENOVO\\Desktop\\air-pollution-data\\node_modules\\jest-circus\\build\\run.js:155:3)
    at _runTestsForDescribeBlock (C:\\Users\\LENOVO\\Desktop\\air-pollution-data\\node_modules\\jest-circus\\build\\run.js:66:9)
    at _runTestsForDescribeBlock (C:\\Users\\LENOVO\\Desktop\\air-pollution-data\\node_modules\\jest-circus\\build\\run.js:60:9)
    at run (C:\\Users\\LENOVO\\Desktop\\air-pollution-data\\node_modules\\jest-circus\\build\\run.js:25:3)
    at runAndTransformResultsToJestFormat (C:\\Users\\LENOVO\\Desktop\\air-pollution-data\\node_modules\\jest-circus\\build\\legacy-code-todo-rewrite\\jestAdapterInit.js:170:21)
    at jestAdapter (C:\\Users\\LENOVO\\Desktop\\air-pollution-data\\node_modules\\jest-circus\\build\\legacy-code-todo-rewrite\\jestAdapter.js:82:19)
    at runTestInternal (C:\\Users\\LENOVO\\Desktop\\air-pollution-data\\node_modules\\jest-runner\\build\\runTest.js:389:16)
    at runTest (C:\\Users\\LENOVO\\Desktop\\air-pollution-data\\node_modules\\jest-runner\\build\\runTest.js:475:34)
    at Object.worker (C:\\Users\\LENOVO\\Desktop\\air-pollution-data\\node_modules\\jest-runner\\build\\testWorker.js:133:12)
  </pre>
  <p>
    üíø Hey developer üëã
  </p>
  <p>
    You can provide a way better UX than this when your app throws errors by providing your own 
    <code
      style="padding: 2px 4px; background-color: rgba(200, 200, 200, 0.5);"
    >
      ErrorBoundary
    </code>
     or 
    <code
      style="padding: 2px 4px; background-color: rgba(200, 200, 200, 0.5);"
    >
      errorElement
    </code>
     prop on your route.
  </p>
</DocumentFragment>
`;

exports[`Component testing Navbar component should render correctly 1`] = `
<DocumentFragment>
  <nav
    class="navbar navbar-expand-lg navbar-expand-sm mx-2 bg-dark"
  >
    <h3
      class="text-light m-0"
    >
      Air pollution Data
    </h3>
    <ul
      class="navbar-nav ms-auto d-flex flex-row"
    >
      <li
        class="nav-item"
      >
        <a
          aria-current="page"
          class="nav-link active"
          href="/"
        >
          <img
            alt="mic"
            height="30px"
            src="mic.png"
          />
        </a>
      </li>
      <li
        class="nav-item"
      >
        <a
          aria-current="page"
          class="nav-link active"
          href="/"
        >
          <img
            alt="setting"
            src="settings.png"
          />
        </a>
      </li>
    </ul>
  </nav>
</DocumentFragment>
`;
